Implementation of replaceOrCreate and replace

Amir-61 opened the pull request:
See also https://github.com/strongloop-internal/scrum-loopback/issues/621

##### Added functionality:

* Initial implemetation for replaceAttributes method
* Initial implemetation for replaceOrCreate

##### Other notes:

* These changes are related to the changes in the pull requests under `loopback-connector` and `loopback-connector-mysql`
* It seems test cases cannot be run on windows machines; I need to dig into it.


Connect to a spike task: https://github.com/strongloop-internal/scrum-loopback/issues/621

bajtos commented on lib/connectors/memory.js:
```diff
@@ -252,6 +252,18 @@ Memory.prototype.save = function save(model, data, options, callback) {
+  var id = this.getIdValue(model, data);
+  var cachedModels = this.collection(model);
+  var modelData = cachedModels && this.collection(model)[id];
+  modelData = modelData && deserialize(modelData);
```
AFAICT, `modelData` is used only to compute `isNewInstance`, in which case I think `deserialize(modelData)` is premature pessimization.

bajtos commented on lib/connectors/memory.js:
```diff
@@ -252,6 +252,18 @@ Memory.prototype.save = function save(model, data, options, callback) {
   });
 };
 
+Memory.prototype.replace = function save(model, data, options, callback) {
```
AFAICT, this method is not used by `lib/dao.js` at all. What is the intended usage?

bajtos replied:
Oh I see, you use it from `replaceAttributes`. Hmm, I find this little bit confusing. Could you perhaps inline this inside `replaceAttributes`?

bajtos commented on lib/connectors/memory.js:
```diff
@@ -679,6 +691,41 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+      throw err;
+    }
+  }
+  this.setIdValue(model, data, id);
```
Please create a copy of `data` before making any modifications, see `updateAttributes` above.

bajtos commented on lib/connectors/memory.js:
```diff
@@ -679,6 +691,41 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+      });
+    } else {
+      self.create(model, data, options, function (err, id) {
+        self.setIdValue(model, data, id);
```
I guess same comment applies here, you should create a clone of `data` before modifying it.

bajtos commented on lib/dao.js:
```diff
@@ -586,6 +586,195 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+
+        var connector = self.getConnector();
+
+        // TODO: validateReplaceOrCreate needs to be set to true by default in model JSON
```
Please remove `validateReplaceOrCreate`, `validateUpdateOrCreate` was added for backwards compatibility only.

bajtos commented on lib/dao.js:
```diff
@@ -586,6 +586,195 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+      var context = {
+        Model: Model,
+        where: ctx.query.where,
+        data: data,
```
Since this operation is modifying the entire instance (instead of a subset of instance properties), all hooks should pass `ctx.instance` instead of `ctx.where` + `ctx.data`.

bajtos commented on lib/dao.js:
```diff
@@ -586,6 +586,195 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+        if (!(data instanceof Model)) {
+          inst = new Model(data, { applyDefaultValues: false });
+        }
+        update = inst.toObject(false);
```
I think this whole block (L658-L664) can be removed, because the hook is manipulating directly the instance, not `data`. Please check the implementation of `create`.

bajtos commented on lib/dao.js:
```diff
@@ -586,6 +586,195 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+        } else {
+          inst.isValid(function(valid) {
+            if (!valid) {
+              if (Model.settings.validateReplaceOrCreate) {
```
Same comment here, just report the error via `cb`.

bajtos commented on lib/dao.js:
```diff
@@ -586,6 +586,195 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+          context = {
+            Model: Model,
+            where: ctx.where,
+            data: update,
```
The "persist" hook always sends `ctx.data` and `ctx.where`, regardless of the operation. So this place should not be changed to `ctx.instance`.

bajtos commented on lib/dao.js:
```diff
@@ -586,6 +586,195 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+          };
+          Model.notifyObserversOf('persist', context, function(err) {
+            if (err) return done(err);
+            if (connector.replaceOrCreate.length === 4) {
```
Since this is a new connector operation and we don't have to preserve backwards compatibility, you can always assume the connector method takes 4 arguments and get rid of this `if` check.

bajtos commented on lib/dao.js:
```diff
@@ -2325,6 +2514,20 @@ DataAccessObject.prototype.updateAttribute = function updateAttribute(name, valu
+  var data = {};
+  data[name] = value;
+  return this.replaceAttributes(data, options, cb);
+};
```
Ooops, this will delete all other existing attributes and modify the object to have only single attribute provided in args. That's almost certainly not wanted, please delete `replaceAttribute` completely.

bajtos commented on lib/dao.js:
```diff
@@ -2356,6 +2559,200 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+  var context = {
+    Model: Model,
+    where: byIdQuery(Model, getIdValue(Model, inst)).where,
+    data: data,
```
Again, because this operation modified the instance as whole, you should pass `ctx.instance` instead of `ctx.data` + `ctx.where` to "before save" hook.

bajtos commented on lib/dao.js:
```diff
@@ -2356,6 +2559,200 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+    if (err) return cb(err);
+    data = ctx.data;
+
+    if (strict && !allowExtendedOperators) {
```
I think "replace" should not support extended operators. Please remove this from the initial PR.

bajtos commented on lib/dao.js:
```diff
@@ -2356,6 +2559,200 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+              && typeof typedData[key].toObject === 'function') {
+              typedData[key] = typedData[key].toObject();
+            }
+          }
```
I assume this was copied from `updateAttributes`. Could you please extract this code to a shared function called from both places?

bajtos commented on lib/dao.js:
```diff
@@ -2356,6 +2559,200 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+              // with the changes made through persist/loaded hooks. To preserve
+              // backwards compatibility, we introduced a new setting updateOnLoad,
+              // which if set, will apply these changes to the model instance too.
+              if(Model.settings.updateOnLoad) {
```
Please remove this flag, there is no backwards compatibility to preserve here.

bajtos commented on lib/dao.js:
```diff
@@ -2356,6 +2559,200 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+            options: options
+          };
+          Model.notifyObserversOf('persist', ctx, function(err) {
+            if (connector.replaceAttributes.length === 5) {
```
There are no 4-arg connectors to support, always assume 5-arg version.

bajtos commented:
Hi @Amir-61, this is fantastic as far as the initial implementation goes. I left few comments to address, you will have to add extensive test coverage before this can be landed.

@fabien @raymondfeng could you PTAL?

bajtos commented:
I'd like to discuss the naming scheme for these new methods (replaceOrCreate, replaceAttributes). I feel that `replaceAttributes` is not describing the intent precisely, and the difference between `replaceAttributes` and `updateAttributes` may not be clear enough.

At the moment, we have (among others):
 - `MyModel.create`
 - `MyModel.updateOrCreate`
 - `myModelInstance.updateAttributes`
 - `myModelInstance.updateAttribute` (a single attribute)

I am proposing to come up with new names while we are at adding support for REPLACE operations:
 - `MyModel.create` (remains same)
 - `MyModel.patchOrCreate` (with an alias `MyModel.updateOrCreate` for backwards-compatibility)
 - `myModelInstance.patchAttributes` (with an alias `myModelInstance.updateAttributes` for backwards-compatibility)
 -  `myModelInstance.updateAttribute`
 - `MyModel.replaceOrCreate`
 - `myModelInstance.replace` or perhaps `myModelInstance.updateTo`, `myModelInstance.updateAllAttributes` or even `myModelInstance.setTo`

Thoughts? 
@raymondfeng @ritch @superkhau @strongloop/loopback-dev @fabien @STRML @clarkorz

superkhau commented:
@bajtos How does replace differ from update?

Amir-61 commented:
@superkhau Sorry do you mean in terms of functionality? 

If you have this model instance `{ name: 'foo', surname: 'bar', age: 20}`
`replaceAttribute('name', 'baz', cb} -> { name: 'baz'}`

while:

`updateAttribute('name', 'baz', cb} -> { name: 'baz', surname: 'bar', age: 20}`

and 

If you have this model instance `{ name: 'foo', surname: 'bar', age: 20}`
`replaceAttributes({'name': 'baz', 'age': 21}, cb) -> { name: 'baz', age: 21}`

While:

`updateAttributes({'name': 'baz', 'age': 21}, cb) -> { name: 'baz', surname: 'bar', age: 21}`


I added the stroy here too: https://github.com/strongloop/loopback/issues/1876

davidcheung commented:
I believe the standard is moving towards using `PATCH` for update attributes (editing just the key value pairs sent), and `PUT` for replacing the whole object with the received payload.

But if user upgraded without knowing it would likely erase their data. :fearful: 

Came across this link that talks about it http://restful-api-design.readthedocs.org/en/latest/methods.html#patch-vs-put

superkhau commented:
>I believe the standard is moving towards using PATCH for update attributes (editing just the key value pairs sent), and PUT for replacing the whole object with the received payload.

TY for the link.

>But if user upgraded without knowing it would likely erase their data. 

Haha yeah, sounds dangerous.

bajtos commented:
> But if user upgraded without knowing it would likely erase their data.

Yes, that's the concern I share too. We must clearly describe this change in release notes and instructions for upgrading.

If we think it's still not good enough, then I guess we will have to implement a feature flag for this and throw an error if the user does not explicitly tell us how to behave. Then in the next major version 4.0, we can make this flag default to "PUT replaces whole object".

Thoughts?

raymondfeng commented:
I am out of the office until 01/04/2016.

Happy holidays!


Note: This is an automated response to your message  "Re:
[loopback-datasource-juggler] Implement replaceOrCreate and
replaceAttributes (#788)" sent on 12/21/15 2:29:08 PM.

This is the only notification you will receive while this person is away.

superkhau commented:
>We must clearly describe this change in release notes and instructions for upgrading.

The trouble is with the people who aren't gonna read the docs (my guess is a large portion of the users) and they will be vocal about it (in triage channels).

>If we think it's still not good enough, then I guess we will have to implement a feature flag for this and throw an error if the user does not explicitly tell us how to behave. Then in the next major version 4.0, we can make this flag default to "PUT replaces whole object".

I agree with this approach, or even without a feature flag we can add a setting in `config.json` (ie. "{"rest": {"usePatch": false}}), then explain the config setting in the docs (which they can look up). We default to false and then force them to read the docs to set true (otherwise the behaviour remains the same, backwards compatible). Then in some major version, we announce the default is going to change to true so too bad, but if you need old behaviour, you have the option to change it back.

bajtos commented:
@Amir-61 I am reassigning this PR back to you, so that it stays out of my list of PRs waiting for my review. Please assign it back to me once you have addressed the comments above and the PR is ready for another round of review.

bajtos commented:
@raymondfeng could you please let me know what do you think about my earlier comment (https://github.com/strongloop/loopback-datasource-juggler/pull/788#issuecomment-164428158) proposing rename of methods?

> I'd like to discuss the naming scheme for these new methods (replaceOrCreate, replaceAttributes). I feel that `replaceAttributes` is not describing the intent precisely, and the difference between `replaceAttributes` and `updateAttributes` may not be clear enough.

> At the moment, we have (among others):
 - `MyModel.create`
 - `MyModel.updateOrCreate`
 - `myModelInstance.updateAttributes`
 - `myModelInstance.updateAttribute` (a single attribute)

> I am proposing to come up with new names while we are at adding support for REPLACE operations:
 - `MyModel.create` (remains same)
 - `MyModel.patchOrCreate` (with an alias `MyModel.updateOrCreate` for backwards-compatibility)
 - `myModelInstance.patchAttributes` (with an alias `myModelInstance.updateAttributes` for backwards-compatibility)
 -  `myModelInstance.updateAttribute`
 - `MyModel.replaceOrCreate`
 - `myModelInstance.replace` or perhaps one of 
  - `myModelInstance.updateTo`
  - `myModelInstance.updateAllAttributes` 
  - `myModelInstance.setTo`

raymondfeng commented on lib/dao.js:
```diff
@@ -2356,6 +2559,200 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+ * @param {Function} cb Callback function called with (err, instance)
+ */
+
+DataAccessObject.prototype.replaceAttributes = function replaceAttributes(data, options, cb) {
```
Can we rename it to `replace` as `replaceAttributes` hints only the attrs will be replaced?

superkhau replied:
+1

Amir-61 commented:
Hi @raymondfeng,

Could you also please review this patch as well? This patch is not done yet; Miroslav did one review and I applied all of his feedback; I really appreciate it if you review it at this point and advise me if I'm on the right direction or not as this patch is getting bigger and bigger.

Notes:
+This patch requires extensive testing (If you have any comment please let me know)

+Since I started this patch some bugs have been solved in the master branch and when I merge master to my current branch it solves some failures; could you please let me know if it is fine I merge master into this branch (`replace`), from which I'm sending the pull request? Or I need to merge it just when I'm done with this pull request?

Any feedback is really appreciated in this patch.

Thanks again :-)

Amir-61 commented:
Hey @raymondfeng,

I just wanted to let you know this patch introduces some sort of refactoring for `updateAttributes` as well to use as much code re-usability as possible. I just went through the diff and I found it not easy to see the flow; I guess you may find it easier if you clone it and see the whole file...

Thanks!

Amir-61 commented on lib/connectors/memory.js:
```diff
@@ -148,11 +148,13 @@ Memory.prototype.saveToFile = function (result, callback) {
-          ids: self.ids,
-          models: self.cache
+          ids: ids,
+          models: cache
```
If I use the following block of code, self.ids and self.cache does not reflect the latest cache and ids (It misses the last item)

```
        var data = JSON.stringify({
          ids: self.ids,
          models: self.cache
        }, null, '  ');
```
However the following code works; cache and ids reflect the updated cache and ids:

```
      var cache = self.cache;
      var ids = self.ids;
      self.writeQueue = async.queue(function (task, cb) {
        // Flush out the models/ids
        var data = JSON.stringify({
          ids: ids,
          models: cache
        }, null, '  ');

        fs.writeFile(self.settings.file, data, function (err) {
          cb(err);
          task.callback && task.callback(err, task.data);
        });
      }, 1);
```

raymondfeng replied:
The 1st case references `self` while the 2nd one keeps pointers to `self.cache` and `self.ids`. If `self.cache` or `self.ids` has not been reassigned, the result should be the same.

superkhau replied:
@raymondfeng +1, I don't see the point of the reassignment.

bajtos replied:
This looks like a possible problem (a race condition?) in the way how we schedule writes via `writeQueue` (`async.queue`), we must pay extra attention before modifying this part of code.

@Amir-61 what is the test case that fails without your change? I would prefer to move this fix to a standalone pull request, so that we can think it through in isolation. There are already too many other changes in this PR.

Amir-61 replied:
@bajtos Yes too many things are happening in this PR; probbaly I will move it to a standalone PR

Amir-61 replied:
I spent some time figuring out when this happens; it seems the first time you run the app and the db file does not exist, it does create the file and it saves data in your file as expected; but if you stop the server and run it, it sometimes does not write in the file. Please let me do more investigation on this to make sure...

Amir-61 replied:
@bajtos 

OK that's my conclusion after lots of investigation; I have always had this problem in memory connector not only in my feature branch, but also in master branch... The problem that I experience is you can `create` a record using memory and saving in a file for the first time you run the app and it creates the file and saves your record(s) in the file. but if you restart the server and use the same file for your DB and add another record with another id, it does not save it in the file anymore.

As you see I see different data inside and outside of `async.queue`
```
      console.log('> cache:', self.cache, 'ids:', self.ids); // It shows the updated data
      self.writeQueue = async.queue(function (task, cb) {
        console.log('> cache:', self.cache, 'ids:', self.ids); // It shows the old data
        var data = JSON.stringify({
          ids: self.ids,
          models: self.cache
        }, null, '  ');

        fs.writeFile(self.settings.file, data, function (err) {
          cb(err);
          task.callback && task.callback(err, task.data);
        });
      }, 1);
```

However, although this does not work for me on my work PC, I used the same juggler version or even this feature branch in juggler in two other computers and it works like a charm. It is worth mentioning we all use  node 4.2.2.

In terms of testing, first of all I'm not sue how I can reproduce the issue, because it happens just when you restart the server, secondly it seems it passes on other computers except my work PC...

Amir-61 replied:
Sorry please ignore my previous comment...

Amir-61 commented on lib/connectors/memory.js:
```diff
@@ -679,6 +681,52 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+  }
+
+  // Do not modify the data object passed in arguments
+  // data = Object.create(data);
```
After this line of code `data = Object.create(data);` data will be an empty object; not sure why!

raymondfeng replied:
```
var y = Object.create({x:1});
console.log(y); // {}
console.log(y.x); // 1 
```
Please note the {x: 1} will be set as the prototype of y. As a result, the x property is not owned by y but it's on the prototype chain.

bajtos replied:
Take a look at L712/714 - it uses the same trick. I don't see why this should not work here.

Amir-61 commented on lib/connectors/memory.js:
```diff
@@ -679,6 +681,52 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+      });
+    }
+  });
+};
```
This `replaceOrCreate` block of code is prone to race condition here and it is not atomic; to solve this I need to have my other PR (strongloop/loopback-datasource-juggler/pull/815) merged to master branch and then merge it to my current branch as well; since, in that patch I implemented synchronous `_findAllSkippingIncludes` which can be used  instead of `exists` and synchronous `_create` can be used inseatd of `create`...

bajtos replied:
AFAICT, #815 has been already merged, could you please rework this implementation into an atomic one?

Amir-61 replied:
@bajtos Sure and you are right I merged [#815](https://github.com/strongloop/loopback-datasource-juggler/pull/815) into `master`! But now my question is should I merge [#815](https://github.com/strongloop/loopback-datasource-juggler/pull/815) into my current branch or `master` into my current branch? The advantegous of the latter is if there is any merge conflicts I can solve them in this feature branch as well?

raymondfeng commented on lib/dao.js:
```diff
@@ -2356,6 +2566,184 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+ * @param {Function} cb Callback function called with (err, instance)
+ */
+
+DataAccessObject.prototype.replace = function replace(data, options, cb) {
```
I'm debating if `replace` should be a prototype method. Maybe a static method such as `replaceById` is more consistent with `updateById` and `destroyById`. If that's agreed, I prefer to use `replace` as an alias to `replaceAll`.

bajtos replied:
@Amir-61 please rename to `DataAccessObject.replaceById` (no `prototype`) as proposed by Raymond.

Amir-61 replied:
@bajtos Sorry I'm not very clear what is the proposed approach:

Could you please confim if I understood you clearly:
You are proposing to change the change this method to the following:

Please note I added `id` to the signature of `replaceById` method and it is not a prototype method anymore.
```
DataAccessObject.replaceById = function(id, data, options, cb) {
   ...
}
```
My understanding is that still we should have a replace prototype method like the following:

```
DataAccessObject.prototype.replace = function replace(data, options, cb) {
  var Model = this.constructor;
  return Model.replaceById(getIdValue(Model, data), data, options, cb);
};
```

Did I understand it correctly?

bajtos replied:
> `DataAccessObject.replaceById = function(id, data, options, cb)`

Exactly.

> `DataAccessObject.prototype.replace = function replace(data, options, cb) {`

The signature is OK, but the name should be changed to `replaceAttributes` to match `updateAttributes` (or `patchAttributes`). The name is not perfect, but I don't have any counter-proposal that would be significantly better, so let's stick with `replaceAttributes` for consistency.

Amir-61 commented on test/persistence-hooks.suite.js:
```diff
@@ -1816,6 +1916,147 @@ module.exports = function(dataSource, should) {
       });
     });
 
+    describe('PersistedModel.replaceOrCreate', function() {
```
Ummm It seems most of the test cases for hooks are valid for replaceOrCreate as well...

bajtos replied:
Yes, that's correct. Please consider changing the copied test cases so that they would fail if run against `updateOrCreate` instead of `replaceOrCreate`.

bajtos commented:
@raymondfeng let's discuss the naming of the new methods. You wrote:

> I'm debating if `replace` should be a prototype method. Maybe a static method such as `replaceById` is more consistent with `updateById` and `destroyById`. If that's agreed, I prefer to use `replace` as an alias to `replaceAll`.

My initial thinking was to make `prototype.replace` similar to existing `prototype.updateAttributes` and eventually use it inside `instance.save`. Note we don't have `updateById` yet (I think `patchById` would be a better name BTW).

OTOH, `replaceById` elegantly solves the naming problem I have outlined before, and it also allows a more efficient implementation (no need to fetch the instance when we are going to replace all attributes anyways). I have a small concern about ACLs, whether the current mechanism is able extract the `id` information from `replaceById` arguments to apply correct ACL checks. I guess it has to, if `destroyById` is remoted. Besides this, I am ok to run with your proposal. 

I personally dislike `replace` as an alias to `replaceAll` - to me it's not clear whether `replace` is manipulating a single instance or multiple instances. Also note that `replaceAll` is out of scope of this work, or at least that was our plan.

Compiling the discussion above, here is my proposal:

Add the following new methods:
 - `MyModel.replaceOrCreate`
 - `MyModel.replaceById`

Add two new aliases:
 - `MyModel.patchOrCreate` (with an alias `MyModel.updateOrCreate` for backwards-compatibility)
 - `myModelInstance.patchAttributes` (with an alias `myModelInstance.updateAttributes` for backwards-compatibility)

@raymondfeng Do you agree?

Amir-61 commented:
@slnode test please

bajtos commented on lib/connectors/memory.js:
```diff
@@ -723,6 +725,62 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+  var cachedModels = this.collection(model);
+  var modelData = cachedModels && this.collection(model)[id];
+  if (modelData) {
+    var id = this.getIdValue(model, data);
```
No need to create another `id` variable - you already received `id` in function arguments.

bajtos commented on lib/connectors/memory.js:
```diff
@@ -723,6 +725,62 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+      cb(err, self.fromDb(model, data));
+    });
+  } else {
+    cb(new Error('Could not replace. Object with id ' + id + ' does not exist!'));
```
Please use a reverse-logic + early return to keep the code easier to read.

```js
if (!modelData) {
  cb(new Error('...'));
}
```

bajtos commented on lib/connectors/memory.js:
```diff
@@ -723,6 +725,62 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+      return cb(err);
+    } else {
+      throw err;
+    }
```
The code below already requires `cb` to be set, I think we can do the same here.

```js
if (!id) {
  return cb(new Error('...'));
}
```

bajtos commented on lib/connectors/memory.js:
```diff
@@ -723,6 +725,62 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
   }
 };
 
+Memory.prototype.replace = function replace(model, id, data, options, cb) {
```
I find `replace` too vague - is it replacing one or multiple instances? Can we rename this method to `Memory.prototype.replaceById` to match the DAO method exposed on models? Or perhaps `Memory.prototype.replaceInstance`.

Amir-61 replied:
Then I believe I need to change the names in other PRs for connector, mysql and mongodb as well

bajtos commented on lib/connectors/memory.js:
```diff
@@ -723,6 +725,62 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+  var self = this;
+  var idName = self.idNames(model)[0];
+  var idValue = self.getIdValue(model, data);
+  var filter = {'where': {}};
```
I think the quotes are redundant, please remove. `{where: {}}`

bajtos commented on lib/connectors/memory.js:
```diff
@@ -723,6 +725,62 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+    // Calling _createSync to update the collection in a sync way and 
+    // to guarantee to create it in the same turn of even loop
+    return self._createSync(model, data, function(err, id) {
+      if (err) return callback(err);
```
You need to wrap the callback in `process.nextTick` here.

bajtos commented on lib/connectors/memory.js:
```diff
@@ -723,6 +725,62 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+        callback(err, data, { isNewInstance: true });
+      });
+    });
+  } else {
```
I think `else` is redundant here, because the previous block always returns via `return self._createSync();`

bajtos commented on lib/connectors/memory.js:
```diff
@@ -723,6 +725,62 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+    var id = self.getIdValue(model, data);
+    self.collection(model)[id] = serialize(data);
+    self.saveToFile(data, function(err) {
+      callback(err, data, { isNewInstance: false });
```
AFAICT, `replace` calls `self.fromDb(data)` before passing `data` to the callback. I think we should be consistent and either call it everywhere in the new code or not at all.

@raymondfeng I am not familiar with `self.fromDb`, do you happen to remember whether we should or not should apply it on the data passed to the callback?

bajtos commented on lib/connectors/memory.js:
```diff
@@ -723,6 +725,62 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+      if (err) return callback(err);
+      self.saveToFile(id, function(err, id) {
+        self.setIdValue(model, data, id);
+        callback(err, data, { isNewInstance: true });
```
This is another place where we may need to apply `self.fromDb` on `data`.

bajtos commented on lib/dao.js:
```diff
@@ -52,6 +52,45 @@ function getIdValue(m, data) {
+  return typedData;
+}
+
+DataAccessObject.prototype.strictChk = function strictChk(strict, data, inst, cb) {
```
Don't use abbreviations please. 

Also if this is a private method, then it would be better to move it out of `DataAccessObject.prototype`. Otherwise it would end up mixed into Model classes, e.g. `Car.prototype.strictChk`.

bajtos commented on lib/dao.js:
```diff
@@ -52,6 +52,45 @@ function getIdValue(m, data) {
+      inst.__unknownProperties.push(key);
+    }
+  }
+  cb(null, removeUndefined(result));
```
I find it suspicious that we call `removeUndefined` here, it's not what I would expect from a function handling "strict" model. Also when "strict" mode is off, `removeUndefined` is not called at all, which looks like a bug to me. Could you please check? I would expect that `removeUndefined` is either called in all cases, regardless of "strict" value, or never.

@raymondfeng do you happen to know what's the correct behaviour?

Amir-61 replied:
You are right! I guess the right behaviour is to call `removeUndefined` in both cases;

@raymondfeng Please correct me if I'm wrong.

CC: @bajtos 

Thanks!

bajtos commented on lib/dao.js:
```diff
@@ -587,6 +626,177 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+ * Replace or insert a model instance: replace exiting record if one is found, such that parameter `data.id` matches `id` of model instance;
+ * otherwise, insert a new record.
+ *
+ * NOTE: No setters, validations, or hooks are applied when using upsert.
```
This is not correct, we should certainly validate the instance in `replaceOrCreate`, since it is always a full instance (instead of a partial update in case of `updateOrCreate`). I think we should apply setters and hooks too.

@raymondfeng thoughts?

bajtos commented on lib/dao.js:
```diff
@@ -587,6 +626,177 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+    if (typeof data === 'function') {
+      cb = data;
+      data = {};
+    }
```
I find this branch weird. What's the expected outcome of `replaceOrCreate(callback)`? I think it's an invalid usage of the API and we should simply report an error via `assert`.

bajtos commented on lib/dao.js:
```diff
@@ -587,6 +626,177 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+          };
+          Model.notifyObserversOf('persist', context, function(err) {
+            if (err) return done(err);
+            connector.replaceOrCreate(Model.modelName, update, options, done);
```
I think you should use `context.data` instead of `update` to make the following hook work:

```js
Model.observe('persist', function(ctx, next) {
  ctx.data = { newdata: true };
});
```

bajtos commented on lib/dao.js:
```diff
@@ -587,6 +626,177 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+        function done(err, data, info) {
+          var context = {
+            Model: Model,
+            instance: data,
```
Note that `loaded` hook uses `cxt.data`, not `ctx.instance`

bajtos commented on lib/dao.js:
```diff
@@ -587,6 +626,177 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+            Model: Model,
+            instance: data,
+            hookState: ctx.hookState,
+            options: options
```
You should set `ctx.isNewInstance` using the `info` from connector.

bajtos commented on lib/dao.js:
```diff
@@ -587,6 +626,177 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+            }
+            if (err) {
+              cb(err, obj);
+              if(!err) {
```
This will never happen, you are already inside `if (err)` block.

bajtos commented on lib/dao.js:
```diff
@@ -2365,6 +2575,184 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+    if (typeof data === 'function') {
+      cb = data;
+      data = undefined;
+    }
```
As mentioned above, I don't think this API usage (`replace(callback)`) makes sense, no need to support it.

bajtos commented on lib/dao.js:
```diff
@@ -587,6 +626,177 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+
+        var connector = self.getConnector();
+
+        inst.isValid(function(valid) {
```
Please take into account `Model.settings.automaticValidation` (see the usage elsewhere).

bajtos commented on lib/dao.js:
```diff
@@ -2365,6 +2575,184 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+      // update instance's properties    
+      inst.setAttributes(data);
+
+      inst.isValid(function(valid) {
```
Please take into account `Model.settings.automaticValidation` (see the usage elsewhere).

bajtos commented on lib/dao.js:
```diff
@@ -2365,6 +2575,184 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+        }
+
+        inst.trigger('save', function(saveDone) {
+          inst.trigger('update', function(done) {
```
No need to trigger these legacy hooks in the new code.

bajtos commented on lib/dao.js:
```diff
@@ -2365,6 +2575,184 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+
+              var ctx = {
+                Model: Model,
+                instance: context.instance,
```
The "loaded" hook expects raw data in `ctx.data`.

bajtos replied:
Also please add `ctx.isNewInstance:false`

bajtos commented on lib/dao.js:
```diff
@@ -2365,6 +2575,184 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+                if (err) return cb(err);
+
+                inst.__persisted = true;
+                inst.setAttributes(ctx.data);
```
Please ensure there is a test covering this bit - that updates from the "loaded" hook are reflected in the instance returned to the callback.

bajtos commented on test/loopback-dl.test.js:
```diff
@@ -837,6 +837,78 @@ describe('Models attached to a dataSource', function() {
+      });
+    });
+
+    it('should replace with new data', function(done) {
```
What's the difference between this test and the test above? It seems to me this second one is a super-set of the first one, and thus it's enough to keep only this second one.

bajtos commented on lib/dao.js:
```diff
@@ -2365,6 +2575,184 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+  var context = {
+    Model: Model,
+    instance: inst,
+    currentInstance: inst,
```
`currentInstance` is provided only when `ctx` has `data` + `where`. When there is `ctx.instance`, then there is no need for `ctx.currentInstance`.

bajtos commented:
Hi @Amir-61, I have reviewed most of the patch except the persistence-hooks tests and left a bunch of comments above.

raymondfeng commented:
@bajtos I'm fine with the following static methods:

- MyModel.replaceOrCreate
- MyModel.replaceById
- MyModel.patchOrCreate
- MyModel.patchById

I prefer to only expose remote endpoints from the static methods to avoid 2 calls. The instance methods such as myModelInstance.replaceAttributes/myModelInstance.patchAttributes can be left to the Node.js layer.

I'm also wondering if we can consolidate everything into one api with options, for example,

MyModel.updateById(id, update, options, cb)

options:
  - patch
    - true - Patch the instance
    - false - Replace the instance
  - upsert
    - true: Insert if no match 

See https://docs.mongodb.org/manual/reference/method/db.collection.update/

Amir-61 commented on lib/connectors/memory.js:
```diff
@@ -723,6 +725,61 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+  }
+
+  // Do not modify the data object passed in arguments
+  // data = Object.create(data);
```
> Take a look at L712/714 - it uses the same trick. I don't see why this should not work here.

Still I'm not sure why this line needs to be commented out; I'm using exactlty the same trick as [here](https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/connectors/memory.js#L712-L714)

bajtos replied:
Please provide a detailed report on what is failing (and how) when this line is not commented out, so that we can troubleshoot.

Amir-61 commented:
Hi @bajtos,

I applied most of your feedback; would you mind reviewing one more time please?

I appreciate it!

bajtos commented:
>  I'm fine with the following static methods:
> MyModel.replaceOrCreate
> MyModel.replaceById
> MyModel.patchOrCreate
> MyModel.patchById

Great, I agree too. Let's leave the last two out of scope of this PR though, I created #825 for that.

> I prefer to only expose remote endpoints from the static methods to avoid 2 calls. The instance methods such as myModelInstance.replaceAttributes/myModelInstance.patchAttributes can be left to the Node.js layer.

That's out of scope of this PR too, let's discuss in https://github.com/strongloop-internal/scrum-loopback/issues/662.

> I'm also wondering if we can consolidate everything into one api with options, for example,
> MyModel.updateById(id, update, options, cb)
> options:
>
> - patch
>  - true - Patch the instance
>  - false - Replace the instance
> - upsert
>  - true: Insert if no match

I don't think we can do that easily now, I see two major obstacles:

1) We want to expose these four variants on different REST endpoints and strong-remoting does not allow us to specify constant arguments based on HTTP mapping (i.e. `PATCH /:id` should set `options.patch:true` and `options.upsert:false`).

2) The hooks context for update vs. upsert and patch vs. replace has different properties and values, even the control flow is slightly different. My concern is that we will end up with a huge `updateById` method full of switches and exceptions for different combinations of input parameters.

OTOH, I admit the current implementation is far from ideal, as it contains too much duplicated code.

Having said that, both 1) and 2) can be addressed, I just think that it's out of scope of this PR and also feel that we have more important issues to fix first.

@raymondfeng @Amir-61 thoughts?

Amir-61 commented:
@slnode test please

Amir-61 commented on lib/dao.js:
```diff
@@ -2407,6 +2577,190 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+  return Model.replaceById(id, inst, Model, data, options, strict, cb);
+};
+
+DataAccessObject.replaceById = function(id, inst, Model, data, options, strict, cb) {
```
@bajtos: I guess I'm not following what you have in your mind based on [this task](https://github.com/strongloop/loopback-datasource-juggler/issues/825); in other words, I guess I should not include `inst`, `Model` and even `strict` in `replaceById` signature; but if I don't pass them as the method arguments, how would I access them (especially `inst`) in `replaceById` which is not a prototype method? For instance `instance` is required in `before save` in `replaceById`

If that means I need to change the implementations in connectors as well and do it like a query with `id` to replace tuple(s)... that seems a big change... 

Could you please clarify this more?

Thanks :-)

bajtos replied:
@Amir-61 Yes, `replaceById` should be `function(id, data, options, cb)`. 

The trick is to build build a new instance using the provided `data` - take a look at how `DataAccessObject.create` is implemented for example: https://github.com/strongloop/loopback-datasource-juggler/blob/63482ef0619bfff429e807bfa2af6ea8b6c00edc/lib/dao.js#L277-L297

```js
// trimmed-down version to illustrate my point, see DAO.create for the full version
DataAccessObject.replaceById = function(id, data, options, cb) {
  var Model = this;
  var inst = new Model(data);
  // etc.
};
```

bajtos replied:
@Amir-61 using my example code above, you can access `strict` via `var strict = inst.__strict;`

Amir-61 commented:
@bajtos Could you please review this patch and please clarify if [this comment](https://github.com/strongloop/loopback-datasource-juggler/pull/788#discussion_r49480771) means I need to change the implementation of `replace` which is now called `replaceById` in all connectors (Mysql, MongoDB, Memory and connectors) do it like a query with `id` to replace tuple(s)... ? If that's the case that seems a big change...

Please see [this comment](https://github.com/strongloop/loopback-datasource-juggler/pull/788/files#r50790824)

Thanks :-)

Amir-61 commented:
@slnode test please

Amir-61 commented:
@slnode test please

rmg commented:
@slnode test please

Amir-61 commented on test/persistence-hooks.suite.js:
```diff
@@ -1828,6 +1912,109 @@ module.exports = function(dataSource, should) {
+          }));
+          done();
+        });
+      });
```
> Yes, that's correct. Please consider changing the copied test cases so that they would fail if run against updateOrCreate instead of replaceOrCreate.
TBH, this is the only test which fails against `updateOrCreate` but passes for `replaceOrCreate`.

bajtos replied:
Ah, I think I was too cautious when I wrote that original comment, it's ok to keep the hooks tests as you have them now.

Amir-61 commented on lib/dao.js:
```diff
@@ -2407,6 +2577,201 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+  var Model = this.constructor;
+  var id = getIdValue(this.constructor, this);
+  var idValue = idName(this.constructor)
+  data[idValue] = id;
```
It may not look neat; I added id field and its value to data, because in line 2602 data should have an `id` when creating instance:
```
var inst = new Model(data);
```
And then deleted the id field from data on line 2610:
```
delete data[idValue];
```

So it is probably ugly!?

bajtos replied:
I think you need to set it inside `replaceById`, not only in `prototype.replaceAttributes`
at least if the following is a correct usage:

 ```
Customer.replaceById(123, { name: 'Miroslav' });
```

if ^^^ is not a valid usage (i.e. we require that data must contain the id field too)
then we must implement a validation and return an error when `data.id !== id`
(of course, use `idName` instead of `id` as the key)
thoughts?

Amir-61 commented:
@bajtos Could you please review it?

bajtos commented on test/loopback-dl.test.js:
```diff
@@ -837,6 +837,81 @@ describe('Models attached to a dataSource', function() {
+        post.replaceAttributes({title: 'b'}, function(err, p) {
+          if (err) return done(err);
+          p.id.should.equal(post.id);
+          should.not.exist(p._id);
```
Please use `p.should.not.have.property('_id')`, as it produces much more helpful error messages on failure. Apply this change to all new test code please.

Amir-61 replied:
:+1:

bajtos commented on test/persistence-hooks.suite.js:
```diff
@@ -1374,6 +1374,90 @@ module.exports = function(dataSource, should) {
       });
     });
+    
+    describe('PersistedModel.prototype.replace', function() {
```
`describe('PersistedModel.prototype.replaceAttributes'

bajtos commented on test/persistence-hooks.suite.js:
```diff
@@ -1374,6 +1374,90 @@ module.exports = function(dataSource, should) {
+          observedContexts.should.eql(aTestModelCtx({
+            where: { id: existingInstance.id },
+            data: { name: 'replacedName' },
+            currentInstance: {
```
`replaceAttributes` should provide `ctx.instance` instead of `where` + `data` + `currentInstance`.

Amir-61 replied:
You said not for `persist` though. I believe in [this comment](https://github.com/strongloop/loopback-datasource-juggler/pull/788#discussion-diff-47087458)

bajtos commented:
@Amir-61 please add more tests to `test/persistence-hooks.suite.js` so that we have a test verifying the content of `context` object for each hook triggered (`access`, `persist`, `before save`, `after save`, `loaded`); add tests checking hooks invoked by `replaceById`.

I am not sure how important it's to test hooks invoked by `prototype.replaceAttributes`, considering that it's just a thin wrapper for `replaceById`. You can keep them or remove (rework to `replaceById`?), I'll leave the decision up to you.

bajtos commented on lib/connectors/memory.js:
```diff
@@ -723,6 +723,67 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+  var cachedModels = this.collection(model);
+  var modelData = cachedModels && this.collection(model)[id];
+  if (!modelData) {
+    return cb(new Error('Could not replace. Object with id ' + id + ' does not exist!'));
```
I think this should be wrapped by `process.nextTick`.

bajtos commented on lib/connectors/memory.js:
```diff
@@ -723,6 +723,67 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+  var self = this;
+  if (!id) {
+    var err = new Error('You must provide an id when replacing!');
+    return cb(err);
```
ditto: I think this should be wrapped by `process.nextTick`.

bajtos commented on lib/connectors/memory.js:
```diff
@@ -723,6 +723,67 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
+        return process.nextTick(function() {
+          cb(err);
+        });
+      };
```
Would you mind compressing this block to a single line?

```js
if (err) return process.nextTick(function() { cb(err); });
```

Or at least

```js
if (err) {
  return process.nextTick(function() { cb(err); });
}
```

The current version has too low signal/noise ratio for me.

bajtos commented on lib/dao.js:
```diff
@@ -629,6 +629,176 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+
+        inst.isValid(function(valid) {
+          if (!valid) {
+              return cb(new ValidationError(inst), inst);
```
Indentation is off.

I would personally put `return` on the same line as `if`:

```js
if (!valid) return cb(new ValidationError(inst), inst);
```

bajtos commented on lib/dao.js:
```diff
@@ -629,6 +629,176 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+            Model: Model,
+            where: where,
+            data: update,
+            currentInstance: inst,
```
As commented in tests, you should provide `instance: inst` instead of `where+data+currentInstance`.

Amir-61 replied:
You said not for `persist` though I believe

bajtos replied:
Of course, you are right! I must got confused, this patch is really huge. Sorry for that.

bajtos commented on lib/dao.js:
```diff
@@ -2407,6 +2577,201 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+    var IdName = idNames[i];
+    if (data[IdName] !== undefined && !idEquals(data[IdName], inst[IdName])) {
+      var err = new Error('id property (' + IdName + ') ' +
+              'cannot be updated from ' + inst[IdName] + ' to ' + data[IdName]);
```
This will never happen, because `inst` is initialized from `data`, thus `data[IdName]` is always the same as `inst[IdName]`.

I think you need to check that the `id` argument matches the id property in `data` instead.

bajtos commented:
@Amir-61 I think `replaceOrCreate` should call `applyStrictCheck` too.

bajtos commented on lib/dao.js:
```diff
@@ -2407,6 +2577,201 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+        // Convert idNames to an object since when we check ids
+        // the complexity would be O(1) in Object as opposed to
+        // an array which would be O(N)
+        // Please feel free to let me know if you dont like the idea
```
Considering that the length of `idNames` array is typically `1`, and almost always less than five, I think this is premature optimisation.

bajtos commented on lib/dao.js:
```diff
@@ -2407,6 +2577,201 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+          var p = propKeys[i];
+          if (idNames[p])
+            continue;
+          inst[p] = null;
```
Are the connectors prepared for the fact that they will receive data without `id` fields filled in, i.e. they have to add them explicitely?

bajtos replied:
I see, the memory connector handles this, you can ignore my comment.

bajtos commented:
@Amir-61 One more thing: we need to add tests to the suite run by all connectors, these new tests should be skipped if the connector does not implement REPLACE yet. Please add these new tests to `test/manipulation.test.js`.

```js
if (!getSchema().connector.replaceById) {
  describe.skip('replaceById - not implemented');
} else {
  describe('replaceById', function() {
    // ...
  });
}

// similarly for `replaceOrCreate`
```

Amir-61 commented on lib/dao.js:
```diff
@@ -2407,6 +2574,197 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+  if (connectionPromise) {
+    return connectionPromise;
+  }
+  var PKName = idName(this);
```
I called it `PKName` instead of `idName`, because the variable name conflicts with the function name...Although I believe `PKName` may not be good either; any better name here?

bajtos replied:
What is used in other places? This is not a big deal, I would just perhaps start with lower-case letter, e.g. `pkName`. Upper-case letters are usually used for constructors.

Amir-61 commented on lib/dao.js:
```diff
@@ -2407,6 +2574,197 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+  Model = this.lookupModel(data); // data-specific
+  if (Model !== inst.constructor) inst = new Model(data);
+  var strict = inst.__strict;
+  // delete data[PKName];
```
In terms of expected behavior still not sure whether I need to delete `id` after creating instance or not; please note I'm using this `data` for `data` in `persist` hook which you asked me in [this comment](https://github.com/strongloop/loopback-datasource-juggler/pull/788#discussion-diff-47087458)

So now that I'm not deleting `id` from `data`in `data` from `persist` I expect `id` as well; please see [here](https://github.com/strongloop/loopback-datasource-juggler/pull/788/files#diff-145ab49855b38984ba190919e6e7397dR1412)

bajtos replied:
I don't have a strong opinion here, let's keep `id` in `data`.

Amir-61 commented on lib/dao.js:
```diff
@@ -2407,6 +2575,183 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+      function callConnector() {
+        var idNames = Model.definition.idNames();
+        var propKeys = Object.keys(Model.definition.properties);
+        var nonIdsPropKeys = lodash.difference(propKeys, idNames);
```
Not sure if I'm allowed to use `lodash` or not?

bajtos replied:
lodash adds quite a lot to the size of browser bundle. It's better to avoid it.

Amir-61 commented on test/loopback-dl.test.js:
```diff
@@ -837,6 +837,83 @@ describe('Models attached to a dataSource', function() {
+            p.should.not.have.property('_id');
+            p.title.should.equal('b');
+            // should.equal(p.content, undefined);
+            // should.equal(p.comments, undefined);
```
If the value of a field is not set, `findById` returns `undefined`. Just curious which check is prefered:
1:
```
should.equal(p.content, undefined);
```
or
2:
```
should.not.exist(p.content);
```

bajtos replied:
I would prefer `p.should.have.property('content', undefined)` or `p.should.not.have.property('content')`, depending on what is actually expected.

Amir-61 commented:
Thanks @bajtos for your feedback! I applied all of your feedback except the following which I will continue working on tomorrow:

TODO:
-`replaceOrCreate` should call `applyStrictCheck` too.
-Write more test cases

Also I have added come comments/questions on the code.

Thanks again!

bajtos commented:
@Amir-61 as I wanted to add code to https://github.com/strongloop/loopback-datasource-juggler/blob/54ca06761063e6cab6bb3cac9b2b0c1a790e69bf/support/describe-operation-hooks.js to describe the new methods, I found that `replaceOrCreate` don't support promises.

```
MyModel.replaceOrCreate({ id: ..., name: ... });

AssertionError: The cb must be passed
    at Function.replaceOrCreate (/Users/bajtos/src/loopback/datasource-juggler/lib/dao.js:646:3)
    at replaceOrCreate_create (/Users/bajtos/src/loopback/datasource-juggler/support/describe-operation-hooks.js:65:25)
```

Please add unit-tests to verify that the new methods support promises when invoked both with and without `options` argument:

```js
MyModel.replaceOrCreate(data);
MyModel.replaceOrCreate(data, options);

MyModel.replaceById(id, data);
MyModel.replaceById(id, data, options);

myModel.replaceAttributes(data);
myModel.replaceAttributes(data, options);
```

Amir-61 commented:
> @Amir-61 I think replaceOrCreate should call applyStrictCheck too.
Right! I just added; please note I'm pushing small commits and I DO NOT squash as one single commit ATM, so that your review will be easier by goring through a few small commits...

Amir-61 commented on lib/dao.js:
```diff
@@ -2407,6 +2583,182 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+            data: context.data,
+            isNewInstance:false,
+            options: options
+          };
```
@bajtos This seems wrong! I believe for `loaded` hook there is no need for `instance`; I'm going to delete it in the next commit; Please confirm me. Thanks!

bajtos replied:
You are correct, we don't provide `ctx.instance` for `loaded` hook.

From what I can tell by looking at other DAO methods, the loaded hooks has this context properties:

```js
          Model: ...,
          data: ...,
          isNewInstance: ...,
          hookState: ...,
          options: ...
        };
```

That's why I want to add the new methods to `support/describe-operation-hooks.js` and then check the output table to verify that new context objects are consistent with the existing ones.

Amir-61 commented:
@bajtos,
I have been just adding multiple commits for new codes, I guess that makes the review process easier, so you may just review the added code...Hence, that's why I'm not squashing them as one commit.

Also, it is worth mentioning I may have not followed [standard commit standards](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html), the reason is I will finally squash them as final commit; my apologies for that...

Amir-61 commented on test/persistence-hooks.suite.js:
```diff
@@ -1374,6 +1374,139 @@ module.exports = function(dataSource, should) {
+         * 
+         * [1]:https://github.com/strongloop/loopback-datasource-juggler/pull/788/files#diff-d717a01e40d8164976e4468af1bce663R2678
+         * [2]:https://github.com/strongloop/loopback-datasource-juggler/pull/788/files#diff-d717a01e40d8164976e4468af1bce663R2707
+        */
```
@bajtos: I asked my question in the comment...

bajtos replied:
Yes, it's the core purpose of before save hook to allow observers to modify instance data (properties, attributes). The test case looks legitimate to me and should be passing.

The links `[1]` and `[2]` don't work for me :(

bajtos replied:
@Amir-61 Let's fix this in the next pull request, I'll add a task to the list of left-overs compiled in my other comment.

Amir-61 commented:
Hi @bajtos,

Today I did the followings:

*Applied your feedback

*Called `applyStrictCheck for `replaceOrCreate`

*Added a bunch of test cases for hooks for `replaceAttributes` and  `replaceOrCreate`; they all pass on local machine for both scenarios; that is, if the method is implemented in the connector (atomic) or the workaround version (non-atomic).

*However, it seems for `replaceAttributes` implementation I may have not been in the right track; I added a few comments in the code, could you please clarify that?

*Also, I have had a few changes in `dao.js`, which may break my test cases in my other PRs for connectors, which are `loopback-connector`, `loopback-connector-mysql` and `loopback-connector-mongodb`, so after this PR I'll to get back to those PRs to make sure they pass too; for now I prefer to finish this patch first.

Could you please review this patch again?

I appreciate it in advance :-)

Amir-61 commented:
> so that your review will be easier by goring through a few small commits...

Sorry maybe not :( In a few commits reverted/rectified some previous commits, so you may find it a little bit confusing...

Amir-61 commented:
> @Amir-61 I think replaceOrCreate should call applyStrictCheck too.
Done.

Amir-61 commented on test/persistence-hooks.suite.js:
```diff
@@ -1828,6 +2068,464 @@ module.exports = function(dataSource, should) {
+        to change all of them together to have consistent logic and write test
+        cases for all scenarios. I also prefer to revert back this commit 
+        since this provides just partial support for `option.notify = false`
+        which is just for atomic implementation
```
Hey @bajtos,
Please see the code comments...

Thanks!

bajtos replied:
I see. My concern is about backwards compatibility. Once we publish the implementation that triggers "loaded" hook twice, then we should probably preserve that behaviour until the next major version.

I think I was not clear in my https://github.com/strongloop/loopback-datasource-juggler/pull/788#commitcomment-15763555. I'd like to see the implementation of new replace* methods modified in such way that "loaded" hook is triggered only once, regardless of which connector is in use.

I'm fine with reverting 8b202e8 as you proposed.

Shouldn't this issue be already fixed by using `replaceById` instead of `prototype.replaceAttributes` in the unoptimized path? I mean which DAO methods are triggering those two "loaded" hooks?

AFAICT, `callNonAtomicUpdateOrCreate ` calls these three methods under the hood:

1) `Model.findOne` with `options.notify:false` - no hooks should be triggered
then one of 
2a) `self.replaceById` - this triggers "loaded" hook only once
2b) `obj.save()` - this triggers "loaded" hook only once

That should trigger "loaded" only once. What am I missing?

Amir-61 commented:
@bajtos 

I applied almost all of your feedback and implemented `options.notiofy` for `replaceOrCreate`, although it may make sense to revert it back for the reasons mentioned  in [this comment](https://github.com/strongloop/loopback-datasource-juggler/pull/788#discussion_r51328442).

Thanks!

bajtos commented:
@Amir-61 I have reverted  8b202e8 to make may review easier. I am copying my response to your comment here so that it does not get lost:

I see. My concern is about backwards compatibility. Once we publish the implementation that triggers "loaded" hook twice, then we should probably preserve that behaviour until the next major version.

I think I was not clear in my https://github.com/strongloop/loopback-datasource-juggler/pull/788#commitcomment-15763555. I'd like to see the implementation of new replace* methods modified in such way that "loaded" hook is triggered only once, regardless of which connector is in use.

I'm fine with reverting 8b202e8 as you proposed.

Shouldn't this issue be already fixed by using `replaceById` instead of `prototype.replaceAttributes` in the unoptimized path? I mean which DAO methods are triggering those two "loaded" hooks?

AFAICT, `callNonAtomicUpdateOrCreate ` calls these three methods under the hood:

1) `Model.findOne` with `options.notify:false` - no hooks should be triggered
then one of 
2a) `self.replaceById` - this triggers "loaded" hook only once
2b) `obj.save()` - this triggers "loaded" hook only once

That should trigger "loaded" only once. What am I missing?

bajtos commented on lib/dao.js:
```diff
@@ -629,6 +629,185 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+      Model.notifyObserversOf('before save', context, function(err, ctx) {
+        if (err) return cb(err);
+        if (strict) {
+          applyStrictCheck(Model, strict, data, inst, validateAndCallConnector);
```
"before save" hook may update `ctx.instance`, in which case you need to update `data` too.

The downside is that converting a model instance into a plain data object incurs some cost. Is it possible to change `applyStrictCheck` in such way that would allow us to pass the model instance directly?

bajtos replied:
> "before save" hook may update ctx.instance, in which case you need to update data too.

This should be fixed still in this PR, please start with writing a unit-test that fails with the current implementation and will pass after you fix it.

> The downside is that converting a model instance into a plain data object incurs some cost. Is it possible to change applyStrictCheck in such way that would allow us to pass the model instance directly?

Let's leave this out of scope of this PR, but fix it soon.

bajtos commented on lib/dao.js:
```diff
@@ -629,6 +629,185 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+          delete data[pkName];
+          if (inst) id = inst[pkName];
+        }
+        if (inst) {
```
~I find this check a bit weird, isn't `inst` always set?~

Ah, `inst` is used in the callback arg for `findOne`, could you please rename it to something like `found` or `existingInst` to prevent confusion?

Now that I understand this better, I think my earlier comment about not allowing `if (inst) id = inst[pkName];` to change the `id` value is no longer relevant, I'll delete it.

bajtos commented on lib/dao.js:
```diff
@@ -2407,6 +2586,183 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+  assert((typeof data === 'object') && (data !== null),
+          'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
```
Please move the code massaging input arguments and checking their validity at the top of the method, just after `stillConnecting` call.

bajtos commented on test/persistence-hooks.suite.js:
```diff
@@ -1374,6 +1374,246 @@ module.exports = function(dataSource, should) {
+      
+      it('applies updates from `persist` hook', function(done) {
+        TestModel.observe('persist', pushContextAndNext(function(ctx){
+          ctx.data.extra = 'hook data';
```
Since one of the main purpose of REPLACE is to allow people to delete properties, I think it will be better to change test cases to delete an existing property instead of adding a new one.

In order to get this PR finished soon, let's keep this task out of scope, but make sure to address it in the near future.

bajtos commented on test/persistence-hooks.suite.js:
```diff
@@ -1828,6 +2068,442 @@ module.exports = function(dataSource, should) {
+            } else {
+              triggered.should.eql([
+                'access',
+                'loaded',
```
Is this still true? As I commented above, since `findOne` is called with `options.notify:false`, I would expect it does not trigger the `loaded` hook?

Amir-61 replied:
>As I commented above, since findOne is called with options.notify:false, I would expect it does not trigger the loaded hook?

Very true! You are right thats a bug in `find` which gets called by `findOne; in `find` [irrespective of whether `options.notify`is false or not](https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/dao.js#L1578-L1583) it called [allcb](https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/dao.js#L1495-L1576) where it triggers all hooks; I guess it makes sense if I do not touch it in this PR... It should be done in another PR... thought?

Amir-61 replied:
Update: Just created [an issue](https://github.com/strongloop/loopback-datasource-juggler/issues/836) for this and I will take care of it in another PR if you agree...

bajtos replied:
OK, but please add a comment here to explain why the loaded hook is triggered and include a link to the new GH issue.

bajtos commented on test/persistence-hooks.suite.js:
```diff
@@ -1828,6 +2068,442 @@ module.exports = function(dataSource, should) {
+          });
+      });
+      
+      it('triggers `before save` hook', function(done) {
```
Please rename to "triggers `before save` hook on create".

bajtos commented on test/persistence-hooks.suite.js:
```diff
@@ -1828,6 +2068,442 @@ module.exports = function(dataSource, should) {
+                  extra: undefined
+                },
+                isNewInstance: false
+              }));
```
Please reduce duplication (the example below may need some tweaks):

```js
var expectedContext = aTestModelCtx({
  instance: {
    id: existingInstance.id,
    name: 'replaced name',
    extra: undefined
  }
}));

if (!dataSource.connector.replaceOrCreate) {
  expectedContext.isNewInstance = false;
}

observedContexts.should.eql(expectedContext);
```

bajtos commented on test/persistence-hooks.suite.js:
```diff
@@ -1828,6 +2068,442 @@ module.exports = function(dataSource, should) {
+                instance: { id: 'new-id', name: 'a name', extra: undefined },
+                isNewInstance: true
+              }));
+            }
```
ditto

bajtos commented on test/persistence-hooks.suite.js:
```diff
@@ -1828,6 +2068,442 @@ module.exports = function(dataSource, should) {
+                  name: 'a name',
+                  extra: undefined
+                }
+              }));
```
This is suspicious. Why atomic `replaceOrCreate` provides `ctx.where` and non-atomic version does not? I would expect consistency here.

bajtos replied:
Oh I see, non-atomic implementation does not provide `ctx.where` because a new instance is being created, so there are not records to match `where` filter.

Can you please still apply the cleanup (add `expectedContext`) and add a comment explaining why `ctx.where` is not set in one of the branches/

bajtos commented on test/persistence-hooks.suite.js:
```diff
@@ -1828,6 +2068,442 @@ module.exports = function(dataSource, should) {
+              // For Unoptimized (non-atomic) connector, the callback
+              // function `pushContextAndNext` is called twice.
+              // As a result, observedContexts returns an array and
+              // NOT a single instance.
```
As commented before, I would like to understand why the "loaded" hook is triggered twice and check if there is a reasonably easy solution that would allow us to have it triggered only once.

Amir-61 replied:
Please see [this comment](https://github.com/strongloop/loopback-datasource-juggler/pull/788#discussion_r51350918)

`loaded` hooks is triggered twice in non-atomic version:

* It gets triggered once by "find()" in this chain: "replaceORCreate()->findOne()->find()", which is a bug but I guess it makes sense if I do not touch it here in this PR.

* Also it gets triggered in "replaceAttributes()" in this chain replaceORCreate()->replaceAttributes()

bajtos replied:
Makes sense. As I commented in the other comment, please include that explanation in a code comment here and include a link to the new GH issue.

Amir-61 replied:
Oh I lost this one; I'll do it now

bajtos commented:
@Amir-61 let's keep the investigation of why "loaded" hook is triggered twice for non-optimized connectors out of scope of this PR, so that we can land this sooner.

Here is the list of things that I would like to get fixed in the next pull request, **outside of the current one**:

MUST HAVE
 - add tests to the suite run by all connectors - https://github.com/strongloop/loopback-datasource-juggler/pull/788#issuecomment-175718192
 - add tests for promise-based usage, add unit-tests for that too - https://github.com/strongloop/loopback-datasource-juggler/pull/788#issuecomment-176075379
 - investigate duplicate call of "loaded" hook - https://github.com/strongloop/loopback-datasource-juggler/pull/788#issuecomment-177134834 and other comments

SHOULD HAVE
 - avoid conversion of `inst` to `data` for `applyStrictCheck` - https://github.com/strongloop/loopback-datasource-juggler/pull/788#discussion_r51343212
 - review & rework hook tests to unset a property instead of adding a new one - https://github.com/strongloop/loopback-datasource-juggler/pull/788#discussion_r51343318

**See the inline comments above for the list of things to fix as part of this pull request**, I hope they should be pretty easy to address.

Amir-61 commented on lib/dao.js:
```diff
@@ -2407,6 +2586,183 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+
+  Model.notifyObserversOf('before save', context, function(err, ctx) {
+    if (err) return cb(err);
+
```
@bajtos Regarding this [comment]: I can do `copyData(ctx.instance, data);` so that I consider instance changes in `before save`, but I'm ignoring `data` in this case if instance has not been changed in `before save`... Any suggestions?

Amir-61 replied:
The same approach; but now not using `lodash` anymore

bajtos replied:
>  solved it using lodash by deep cloning the original instance before triggering before save hook and comparing it after before save hook with new instance, so if it has been changed I update data as well; I guess this is the best approach except the fact that I should definitely avoid using lodash; so I need to refactor it not using lodash... Please let me know if you have any better approach?

Deep-cloning and comparing incurs performance penalty for everyone, which is bad for users that don't install any "before save" hooks at all - they have to pay for something they don't want.

I think we should move the code extracting `data` from `instance` only after the hook was invoked. I'll comment more once I review this code in a bigger picture.

Amir-61 commented:
@slnode test please

Amir-61 commented:
@slnode test please

Amir-61 commented:
@slnode test please

Amir-61 commented on test/loopback-dl.test.js:
```diff
@@ -837,6 +837,81 @@ describe('Models attached to a dataSource', function() {
+      });
+    });
+  });
+  
```
These tests are all covered in `manipulation.test.js`; not sure what kind of tests need to be written in `loopback-dl.test.js`; if any needs to be written in this file...

bajtos replied:
If we have enough tests in `manipulation.test.js` then there is no need to add any tests to `loopback-dl.test.js`.

Amir-61 commented on test/manipulation.test.js:
```diff
@@ -688,6 +688,313 @@ describe('manipulation', function () {
   });
 
+  if (!getSchema().connector.replaceOrCreate) {
+    describe.skip('replaceOrCreate - not implemented', function(done) { done; });
```
So we don't want to have tests for other databases for which we use "non-atomic" database query?

bajtos replied:
Oh, I didn't realise we provide a non-atomic shim for connectors that don't provide an atomic version. However, this shim still depends on the connector to provide `replaceById` implementation, doesn't it?

In that case we should change this condition here - run the tests if the connector provides `replaceById`, skip them otherwise.

Here is my intention behind the original comment: I don't want to break CI builds of all connectors out there (including community ones) when this PR lands. At the same time, I want a notification for connector maintainers that their connector is not supporting all juggler features. `describe.skip` fits that well - `mocha -R spec` will print the test name in a different colour and don't fail the build.

bajtos replied:
Also:

The callback of `describe` does not take any arguments. In fact, you don't have to provide any callback function at all.

```js
describe.skip('replaceAttributes/replaceById - not implemented');
```

bajtos commented on lib/dao.js:
```diff
@@ -2407,6 +2590,196 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+  if (data instanceof Model) {
+    data = data.toObject(false);
+  }
+  data = removeUndefined(data);
```
This is weird. You have already created `inst = new Model(data)`, why are you checking for `data instanceof Model` here? You should get the data for connector later via `inst.toObject(false)`.

bajtos commented on lib/dao.js:
```diff
@@ -2407,6 +2590,196 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+        }
+        data[key] = val;
+      }
+    }
```
Please remove this block and `oInst`, just call `data = inst.toObject(false)`

bajtos replied:
Take a look at how `replaceOrCreate` handles this, I think you can use the same approach here.

bajtos commented on test/manipulation.test.js:
```diff
@@ -688,6 +688,313 @@ describe('manipulation', function () {
+  }
+
+  if (!getSchema().connector.replaceById) {
+    describe.skip('replaceAttributes/replaceById - not implemented', function (done) {
```
The callback of `describe` does not take any arguments. In fact, you don't have to provide any callback function at all.

```js
describe.skip('replaceAttributes/replaceById - not implemented');
```

bajtos commented on test/manipulation.test.js:
```diff
@@ -688,6 +688,313 @@ describe('manipulation', function () {
+        });
+      });
+
+      it('without options on create (promise variant)', function(done) {
```
the test name should read like a sentence, where "it" stands for the value from the describe block. At the moment, you have "it/replaceOrCreate with options on create (promise variant)".

The usual convention is to describe what happens and under what conditions. You have the conditions in the test name, but you are missing what is the expected outcome. 

Let's keep this simple and just prefix your test names with "works", i.e.

```js
it('works without options on create (promise variant)', ...);
```

Please apply this change to other test cases too.

bajtos commented on test/manipulation.test.js:
```diff
@@ -688,6 +688,311 @@ describe('manipulation', function () {
     });
   });
 
+  if (!getSchema().connector.replaceOrCreate) {
```
Based on our discussion, I think this should be `!getSchema().connector.replaceById`.

bajtos commented on test/persistence-hooks.suite.js:
```diff
@@ -1374,6 +1374,248 @@ module.exports = function(dataSource, should) {
       });
     });
+    
+    describe('PersistedModel.prototype.replaceAttributes', function() {
```
This should be skipped when the connector does not implement `replaceById`.

bajtos commented on test/persistence-hooks.suite.js:
```diff
@@ -1828,6 +2070,440 @@ module.exports = function(dataSource, should) {
       });
     });
 
+    describe('PersistedModel.replaceOrCreate', function() {
```
ditto

bajtos commented:
Three more comments, the patch LGTM otherwise. Once you address them, squash all commits into a single one, and ensure down-stream dependencies don't have new test failures caused by this change; then you can land this pull request without further review.

Unfortunately dependant builds are failing now, so we cannot simply check green/red status. Please check the build console outputs and make sure the failures are not caused by your patch.

Amir-61 commented:
@slnode test please

Amir-61 commented:
@slnode test please

Amir-61 commented:
@bajtos 

### 51/63 tests pass:
`Loopback-datasource-juggler` passes on all machines.

### Connectors passing: 
* Loopback-connector-sqlite
* Loopback-connector-redis
* Loopback-connector-oracle
* Loopback-connector-mssql
* Loopback-connector-mongodb

### Connector failing:
* Loopback-connector-mysql: It passes on [all machines except on Ubuntu-5](http://ci.strongloop.com/job/loopback-connector-mysql/1005/console); even for Ubuntu-5 it says:
```  747 passing (7m)
  23 pending
```
But test result is failure! I did not find the actual failure though!

I believe it is safe to merge; I just wait for your last confirmation and then I'll merge it.

>  this patch can be back-ported to 2.x too, right?

I beleive so! Once you confirm the small failure for `loopback-connetcor-mysql` on Ubuntu I will backport it to `2.x` as well!

###### Also Thanks so much for your patience for this patch; I really learned a lot in this patch about operation hooks...

Amir-61 commented:
@slnode test please

rmg commented:
For the Ubuntu 5 build above, that is the second run of the tests that passed. The first run had 3 failures:
```

  640 passing (5m)
  21 pending
  3 failing

  1) Persistence hooks "before each" hook: createTestData:
     Uncaught AssertionError: expected null to exist
      at node_modules/loopback-datasource-juggler/test/persistence-hooks.suite.js:736:24
      at node_modules/loopback-datasource-juggler/lib/dao.js:1629:62
      at node_modules/loopback-datasource-juggler/lib/dao.js:1561:9
      at Object.async.each (node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:153:20)
      at allCb (node_modules/loopback-datasource-juggler/lib/dao.js:1497:13)
      at node_modules/loopback-connector/lib/sql.js:1075:7
      at node_modules/loopback-datasource-juggler/lib/observer.js:166:22
      at doNotify (node_modules/loopback-datasource-juggler/lib/observer.js:93:49)
      at MySQL.ObserverMixin._notifyBaseObservers (node_modules/loopback-datasource-juggler/lib/observer.js:116:5)
      at MySQL.ObserverMixin.notifyObserversOf (node_modules/loopback-datasource-juggler/lib/observer.js:91:8)
      at cbForWork (node_modules/loopback-datasource-juggler/lib/observer.js:156:14)
      at node_modules/loopback-connector/lib/sql.js:417:7
      at handleResponse (lib/mysql.js:9:6483)
      at Query._callback (lib/mysql.js:9:7153)
      at Query.Sequence.end (node_modules/mysql/lib/protocol/sequences/Sequence.js:96:24)
      at Query._handleFinalResultPacket (node_modules/mysql/lib/protocol/sequences/Query.js:144:8)
      at Query.EofPacket (node_modules/mysql/lib/protocol/sequences/Query.js:128:8)
      at Protocol._parsePacket (node_modules/mysql/lib/protocol/Protocol.js:280:23)
      at Parser.write (node_modules/mysql/lib/protocol/Parser.js:73:12)
      at Protocol.write (node_modules/mysql/lib/protocol/Protocol.js:39:16)
      at Socket.<anonymous> (node_modules/mysql/lib/Connection.js:96:28)
      at readableAddChunk (_stream_readable.js:146:16)
      at Socket.Readable.push (_stream_readable.js:110:10)
      at TCP.onread (net.js:523:20)

  2) Persistence hooks PersistedModel.findOrCreate Uncaught error outside test suite:
     Uncaught TypeError: Cannot read property 'currentRetry' of undefined
  

  3) Persistence hooks "before each" hook: setupDatabase for "should not see the uncommitted insert":
     TypeError: Cannot read property 'call' of undefined
```

Amir-61 commented:
@rmg Thanks Ryan for clarification!

Just to make sure I re-ran the tests and all connectors and juggler pass now. I will be merging it now.

Thanks!
